Example notebooks

chaedri opened the pull request:
* Example notebooks for `grass.jupyter`
* `font` parameter for GrassRenderer

You can try the notebooks out in Binder here:
https://mybinder.org/v2/gh/chaedri/grass/Example-Notebooks?urlpath=lab%2Ftree%2Fdoc%2Fnotebooks%2Fgrass_jupyter.ipynb

This PR is part of an ongoing Google Summer of Code project, Improved Integration of GRASS GIS and Jupyter Notebooks. You can find more information [here](https://trac.osgeo.org/grass/wiki/GSoC/2021/JupyterAndGRASS).

wenzeslaus reviewed (changes requested):
non-interactive map display: Let's try to find a positive name for it. Renderer Map Display? Not elegant, but it fits with the class name.

_Color palettes must be set with r.colors and v.colors._ Is this true for vectors? I would love to be wrong here, but isn't more that colors and attributes for vectors are not yet supported?

The sans font piece from the original notebook would be nice to add I think. Here or in a separate PR.

```
os.environ['GRASS_FONT'] = 'sans'
```

Typo:

```
----> 5 ggs.run_command("r.colors", map="elevation", color="grey")
NameError: name 'ggs' is not defined
```

Let's get the test PR done first, so the hydrology notebook works. 

Please, add the link to Binder to the description.

chaedri commented:
> non-interactive map display: Let's try to find a positive name for it. Renderer Map Display? Not elegant, but it fits with the class name.

I went with "Rendering GRASS display"; not sure that makes sense grammatically (I've now thought about it too much) but I was hoping it was slightly more elegant?

> _Color palettes must be set with r.colors and v.colors._ Is this true for vectors? I would love to be wrong here, but isn't more that colors and attributes for vectors are not yet supported?

Now, I'm not sure that this is true for rasters even. It seems to work in some instances but not others. For example, adding `gs.run_command('r.colors', map="elevation", co="grey")` to  the interactive map in `grass_jupyter.ipynb`  doesn't do anything for me. But, in the `viewshed_solar.ipynb`, changing the color in the solar radiation section *does* modify the color on `rad_map`.  I think you are right about 'v.colors' though.

> The sans font piece from the original notebook would be nice to add I think. Here or in a separate PR.
Added it here.

> Typo:
Fixed.

> Please, add the link to Binder to the description.
Added

Two other notes:
* The size of the GrassRenderer image seems disproportionate, like the maps are a squished and there is blank white space on the sides. I can't figure out why this is and I don't think it used to do this. Do you get the same result?
* The `viewshed_solar` notebook is a work in progressed based off the `Try GRASS online` notebook by the same title. I am adding it here just for feedback. There are several map options at the end and I'm not sure what fits best, InteractiveMap or GrassRenderer. Feedback welcome.

veroandreo commented on doc/notebooks/grass_jupyter.ipynb:
```diff
@@ -30,46 +26,45 @@
-   "outputs": [],
-   "source": [
-    "# Set computational region to the study area.\n",
+    "gj.init(\"../../../grassdata\", \"nc_basic_spm_grass7\", \"user1\")\n",
```
It does not work here with this path. Why not the one that was before? If I switch back to that one, it runs fine. Otherwise, errors because it does not find the data

chaedri replied:
Yes, the path should be `../../data/grassdata` - I changed it to test on my local machine before pushing and forgot to change it back. Ooops! It'll be fixed on the next commit.

wenzeslaus replied:
I think I marked the conversation as resolved too soon. The path still needs fixing.

Related to that, I guess another future todo is to fail in gj.init when the mapset does not exist...which brings in another issue and that is a good way of creating a location.

chaedri replied:
Noted. Yes, I fixed it in one commit then changed it back in the next one to test it on my computer and then forgot to change it back again... Oops.

wenzeslaus commented on doc/notebooks/grass_jupyter.ipynb:
`````diff
@@ -30,46 +26,45 @@
+    "\n",
+    "Alternatively, *GRASS* display modules can be called with the `run()` method:\n",
+    "````\n",
+    "m = GrassRenderer()`\n",
`````
Extra backtick at the end of the line.

chaedri replied:
I separated computation and visualization and switch to d_shade - that is a neat module!

wenzeslaus reviewed (changes requested):
The code blocks which combine computation and visualization would be better split into two to allow more experimenting. Similarly, in the grass.jupyter-focused notebook, creation of the renderer/map objects should be separated from adding of layers if possible.

I think the elevation with shades from r.sun image looked better when you showed me before. Maybe? Or maybe d_shade would work for this.

wenzeslaus commented on doc/notebooks/basic_example_grass_jupyter.ipynb:
```diff
@@ -0,0 +1,179 @@
+    "gs.run_command('v.buffer', input=\"boundary_state\", output=\"buffer\", distance=-10000)\n",
+    "gs.parse_command('g.region', vector=\"boundary_state\", flags='pg')\n",
+    "\n",
+    "# Start another GrassRenderer Map\n",
```
Remove Map. Refer to the class by its name. Same above.

wenzeslaus commented on doc/notebooks/basic_example_grass_jupyter.ipynb:
```diff
@@ -0,0 +1,179 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "gs.run_command('v.buffer', input=\"boundary_state\", output=\"buffer\", distance=-10000)\n",
```
This comes from the original notebook, but it would be nice to have consistent quotes used everywhere (that is double quotes).

wenzeslaus commented on doc/notebooks/basic_example_grass_jupyter.ipynb:
```diff
@@ -0,0 +1,179 @@
+    "gs.parse_command('g.region', vector=\"boundary_state\", flags='pg')\n",
+    "\n",
+    "# Start another GrassRenderer Map\n",
+    "m = gj.GrassRenderer()\n",
```
r1 and r2 may work as well although not ideal.

chaedri replied:
I went with fig and fig1 here. I also changed a few of the other `m`s to fig.

chaedri replied:
Inspired by Matplotlib, I changed quite a few of these. I went with `img` (and `img2` in this case) for `GrassRenderer` and `fig` for `InteractiveMap`.

wenzeslaus reviewed (changes requested):
This is definitely turning out nicely.

It is okay for more advanced examples to take longer, but you should mention that in the text (at least for "Compute global (beam+diffuse+refl) radiation for entire day..."

wenzeslaus commented on doc/notebooks/grass_jupyter.ipynb:
```diff
@@ -0,0 +1,238 @@
+    "\n",
+    "As part of Google Summer of Code 2021, we've been working to shorten and simplify the launch of GRASS in Jupyter and imporve the map displays. You can find out more about the project and follow the progress on the [GRASS wiki page](https://trac.osgeo.org/grass/wiki/GSoC/2021/JupyterAndGRASS).\n",
+    "\n",
+    "In addition to simplifying the launch of *GRASS GIS* with `init()`, `grass.jupyter` has two main dislay classes, `GrassRenderer` and `InteractiveMap`. `GrassRenderer` displays PNG images created with the *GRASS GIS* renderer. `InteractiveMap` displays *GRASS GIS* rasters and vectors with [*folium*](http://python-visualization.github.io/folium/), a [*leaflet*](https://leafletjs.com/) library  ...
```
The sentence about GrassRenderer is misleading. It really renders, not just displays. It is called rendered after all. As similar sentence in one of the following cells is much better. However, starting from the current sentence something about "using GRASS rendering engine in the background" might be good.

chaedri replied:
I went with : "Using the GRASS rendering engine in the background, GrassRenderer creates maps as PNG images."

wenzeslaus commented on doc/notebooks/basic_example_grass_jupyter.ipynb:
```diff
@@ -0,0 +1,179 @@
+    "\n",
+    "[<img src=\"../../man/grass_logo.png\" alt=\"GRASS GIS\" style=\"width:200px;\"/>](https://grass.osgeo.org/)\n",
+    "\n",
+    "This is a quick introduction to *GRASS GIS* in a *Jupyter Notebook* using the `grass.jupyter` package and the *Python* scripting language. The `grass.jupyter` package shortens the launch of *GRASS GIS* in *Jupyter Notebook* and provides several useful classes for creating, displaying and saving *GRASS GIS* maps. This notebook can be directly compared with `example_notebook.ipynb` to see how  ...
```
The name of the original notebook is wrong and if you write it as [basic_example.ipynb](basic_example.ipynb), it actually opens the other one.

chaedri replied:
Good catch and good idea!

wenzeslaus commented on doc/notebooks/hydrology.ipynb:
```diff
@@ -0,0 +1,382 @@
+   "source": [
+    "## Watershed Area\n",
+    "\n",
+    "With our watersheds, we can compute some zonal statistics. In this section, we use the `\"count\"` method in `r.stats.zonal` to make a map of watershed area."
```
`count` seems to be sufficient to me.

chaedri commented on doc/notebooks/grass_jupyter.ipynb:
```diff
@@ -0,0 +1,238 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Let's see what is in the example database so we can continue to experiment\n",
```
I deleted it since it's in the basic notebook.

wenzeslaus commented:
Merged, the best way to test Caitlin's GSoC grass.jupyter package is now:

https://mybinder.org/v2/gh/OSGeo/grass/master?urlpath=lab%2Ftree%2Fdoc%2Fnotebooks%2Fgrass_jupyter.ipynb

chaedri commented:
> Merged, the best way to test Caitlin's GSoC grass.jupyter package is now:
> 
> https://mybinder.org/v2/gh/OSGeo/grass/master?urlpath=lab%2Ftree%2Fdoc%2Fnotebooks%2Fgrass_jupyter.ipynb

I think master has been changed to main. If the above link doesn't work, try:

https://mybinder.org/v2/gh/OSGeo/grass/main?urlpath=lab%2Ftree%2Fdoc%2Fnotebooks%2Fgrass_jupyter.ipynb
